### PR TITLE
added a test to check register ops done during a pending poll.

### PR DESCRIPTION
**Pull request by dtacalau**
Signed-off-by: Daniel Tacalau <dst4096@gmail.com>

### CODE REVIEW

**Comment by kleimkuhler** [comment]
I like the idea of this test, but I'm not sure this is guaranteed to test what it's being added for. It is pretty race-y with how it is currently implemented, and it's not guaranteed that when `listener2` is registered, the main thread is blocking in a `Poll:poll`. `expect_events` does multiple iterations of `poll`, and the barriers in their current placement don't necessarily block things where they need to be.

To start, it could be helpful to directly `poll.poll` once, and still kind of *hope* that when the first thread is blocked in that call the second thread tries to register, but again that is not guaranteed.

The guaranteed fix would somehow pass in a `Barrier` to `Poll::poll` and have it block in there, but that is **way** more of a change that I don't think we need to do.

So this isn't really a helpful comment, but maybe if you look again at where the wait points are and try to sync it up a little more?

**Comment by Thomasdezeeuw** [comment]
I agree with @kleimkuhler. To make it a little less race-y, you could call `barrier.wait()` just before a direct call to `register` and `Poll::poll` (with a long-ish timeout, e.g. 5 seconds). It doesn't guarantee that calls will be concurrent, but on local machines with enough cores available it will likely result in errors if the call is not made, hopefully failing the test.

**Comment by dtacalau** [comment]
@Thomasdezeeuw @kleimkuhler I agree with both of you. Try to see this test as a best effort test which should work most of the time.  I could redo it with poll(None) w/o expect_events. Second opinion?

Long story:
Initially I've written it using Poll.poll with timeouts, but then I remembered what Carl said about timeouts and CI, and I've redone it w/o timeouts using Poll.poll(None), but got failures from time to time. Next I've reconsidered, again, and reimplemented it with expect_events. I thought expect_events is best because it is used in lots of tests. 

I've run this test with prints many times and most of the times register was done during the pending poll.  So, to wrap it up, I would not add timeouts because of CI limitations, if current expect_events timeouts, 500ms, are not enough, than 5sec would not cut it most of the times too. Having barrier inside poll would be best, but can't be done atm.

**Comment by Thomasdezeeuw** [comment]
@dtacalau I created https://github.com/dtacalau/mio/pull/1 to simplify this pr.

**Comment by dtacalau** [comment]
> @dtacalau I created [dtacalau#1](https://github.com/dtacalau/mio/pull/1) to simplify this pr.

@Thomasdezeeuw While your test is simpler it doesnt fix the race problems. Both thread will race after the first barrier.wait call. Let me think about it.

**Comment by Thomasdezeeuw** [comment]
> @Thomasdezeeuw While your test is simpler it doesnt fix the race problems. Both thread will race after the first barrier.wait call. Let me think about it.

Isn't that what this test was testing? Or have I misunderstood the purpose of the test?

**Comment by dtacalau** [comment]
> > @Thomasdezeeuw While your test is simpler it doesnt fix the race problems. Both thread will race after the first barrier.wait call. Let me think about it.
> 
> Isn't that what this test was testing? Or have I misunderstood the purpose of the test?

@Thomasdezeeuw What I meant was my test and your test suffer from the same `possible` race condition between poll and register, but, nevermind, I've accepted you PR. 

I've also rebased on master.

**Comment by dtacalau** [comment]
I've opened #1226 to track the CI hanging test

File: tests/poll.rs [hunk 1]
```diff
@@ -310,6 +310,43 @@ fn registry_operations_are_thread_safe() {
...
+
+    let registry = Arc::new(poll.registry().try_clone().unwrap());
+
+    let barrier = Arc::new(Barrier::new(2));
+    let barrier1 = Arc::clone(&barrier);
+
+    let handle1 = thread::spawn(move || {
+        let mut stream = UdpSocket::bind(any_local_address()).unwrap();
+
+        barrier1.wait();
```
Comment (kleimkuhler) [inline tests/poll.rs]: Thoughts on adding some millisecond sleep here? That way both threads unblock at the barrier. The poll thread immediately goes into a 5 second poll, and this thread waits just a little bit so we can get closer to "trying" to register during a poll.

**Comment by dtacalau** [comment]
> LGTM, @kleimkuhler suggestion about a small sleep before registering would make this even better.

I've adedd the timeout before register.

**Comment by dtacalau** [comment]
@Thomasdezeeuw I've reverted your commit https://github.com/tokio-rs/mio/pull/1220/commits/c20132df7e65f0f95125a3786a78fd21ebc6fd6f which was causing all Linux CI to fail

**Comment by dtacalau** [comment]
@kleimkuhler , @Thomasdezeeuw I think this can be merged.

**Comment by kleimkuhler** [comment]
@dtacalau And thanks for taking care of all the reviews on this! I think this test does a good job trying to trigger the test purpose.

**Pull request merged**